test/e2e/no-snat.go

```diff
@@ -0,0 +1,274 @@
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	// . "github.com/onsi/gomega"
+)
+
+const (
+	imageSource = "gcr.io/kube-play-ext/"
+
+	testPodPort  = 8080
+	testPodImage = "no-snat-test-amd64"
+	testPodTag   = ":1.0.0"
+
+	testProxyPort  = 31235 // Firewall rule allows external traffic on ports 30000-32767. I just picked a random one.
+	testProxyImage = "no-snat-test-proxy-amd64"
+	testProxyTag   = ":1.0.0"
+)
+
+var (
+	testPod = v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "no-snat-test",
+			Labels: map[string]string{
+				"no-snat-test": "",
+			},
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "no-snat-test",
+					Image: imageSource + testPodImage + testPodTag,
+					Args:  []string{"--port", strconv.Itoa(testPodPort)},
+					Env: []v1.EnvVar{
+						v1.EnvVar{
+							Name:      "POD_IP",
+							ValueFrom: &v1.EnvVarSource{FieldRef: &v1.ObjectFieldSelector{FieldPath: "status.podIP"}},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testProxyPod = v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "no-snat-test-proxy",
+		},
+		Spec: v1.PodSpec{
+			HostNetwork: true,
+			Containers: []v1.Container{
+				{
+					Name:  "no-snat-test-proxy",
+					Image: imageSource + testProxyImage + testProxyTag,
+					Args:  []string{"--port", strconv.Itoa(testProxyPort)},
+					Ports: []v1.ContainerPort{
+						{
+							ContainerPort: testProxyPort,
+							HostPort:      testProxyPort,
+						},
+					},
+				},
+			},
+		},
+	}
+)
+
+// Produces a pod spec that passes nip as NODE_IP env var using downward API
+func newTestPod(nodename string, nip string) *v1.Pod {
+	pod := testPod
+	node_ip := v1.EnvVar{
+		Name:  "NODE_IP",
+		Value: nip,
+	}
+	pod.Spec.Containers[0].Env = append(pod.Spec.Containers[0].Env, node_ip)
+	pod.Spec.NodeName = nodename
+	return &pod
+}
+
+func newTestProxyPod(nodename string) *v1.Pod {
+	pod := testProxyPod
+	pod.Spec.NodeName = nodename
+	return &pod
+}
+
+func getInternalIP(node *v1.Node) (string, error) {
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == v1.NodeInternalIP {
+			return addr.Address, nil
+		}
+	}
+	return "", fmt.Errorf("did not find InternalIP on Node")
+}
+
+func getExternalIP(node *v1.Node) (string, error) {
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == v1.NodeExternalIP {
+			return addr.Address, nil
+		}
+	}
+	return "", fmt.Errorf("did not find ExternalIP on Node")
+}
+
+func getSchedulable(nodes []v1.Node) (*v1.Node, error) {
+	for _, node := range nodes {
+		if node.Spec.Unschedulable == false {
+			return &node, nil
+		}
+	}
+	return nil, fmt.Errorf("all Nodes were unschedulable")
+}
+
+func checknosnatURL(proxy, pip string, ips []string) string {
+	return "http://" + proxy + "/checknosnat?target=" + pip + "&ips=" + strings.Join(ips, ",")
```

Comment:
nit (feel free to ignore here an elsewhere where I commented similarly):
return fmt.Sprintf("http://%s/checknosnat?target=%s&ips=%s", proxy, pip, strings.Join(ips, ","))